src/ucs/memory/rcache.c

```diff
@@ -1069,18 +1069,23 @@ static ucs_status_t ucs_rcache_global_list_add(ucs_rcache_t *rcache)
     return status;
 }
 
-static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache) {
+static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache)
+{
+    ucs_async_pipe_t pipe;
+
     pthread_mutex_lock(&ucs_rcache_global_context.lock);
+    pipe = ucs_rcache_global_context.pipe;
     ucs_list_del(&rcache->list);
     if (!ucs_list_is_empty(&ucs_rcache_global_context.list)) {
         pthread_mutex_unlock(&ucs_rcache_global_context.lock);
         return;
     }
 
-    ucs_async_remove_handler(ucs_async_pipe_rfd(
-                                     &ucs_rcache_global_context.pipe),
+    pthread_mutex_unlock(&ucs_rcache_global_context.lock);
+    ucs_async_remove_handler(pipe.read_fd,
```

Comment:
1. weird indent
2. i think there is still a race here - need to set ucs_rcache_global_context.pipe to -1 inside the first lock section, instead of taking lock in line 1087